### PR TITLE
Implementation of zone digest (ZONEMD)

### PR DESCRIPTION
**Pull request by madestro**
https://tools.ietf.org/html/draft-ietf-dnsop-dns-zone-digest-14

### CODE REVIEW

**Comment by miekg** [comment]
ah nice. Can you clean up this PR and remove any non go files? Also revert the go.mod updates

**Comment by miekg** [comment]
also needs a bit of tests

**Review by tmthrgd — CHANGES_REQUESTED** [review]
Unrelated changes need to be removed.

File: generate.go [hunk 1]
```diff
@@ -188,7 +188,7 @@ func (r *generateReader) ReadByte() (byte, error) {
 			if errMsg != "" {
 				return 0, r.parseError(errMsg, si+3+sep)
 			}
-			if r.start+offset < 0 || int64(r.end) + int64(offset) > 1<<31-1 {
+			if r.start+offset < 0 || int64(r.end)+int64(offset) > 1<<31-1 {
```
Comment (tmthrgd) [inline generate.go:191]: This change is unrelated.

File: go.mod [hunk 2]
```diff
@@ -3,9 +3,9 @@ module github.com/miekg/dns
 go 1.12
 
 require (
-	golang.org/x/crypto v0.0.0-20191011191535-87dc89f01550
-	golang.org/x/net v0.0.0-20190923162816-aa69164e4478
-	golang.org/x/sync v0.0.0-20190423024810-112230192c58
-	golang.org/x/sys v0.0.0-20190924154521-2837fb4f24fe
-	golang.org/x/tools v0.0.0-20191216052735-49a3e744a425 // indirect
+	golang.org/x/crypto v0.0.0-20200622213623-75b288015ac9
```
Comment (tmthrgd) [inline go.mod:6]: The changes to go.mod and go.sum are unrelated.

File: scan_rr.go [hunk 3]
```diff
@@ -846,6 +846,42 @@ func (rr *CSYNC) parse(c *zlexer, o string) *ParseError {
...
+	if e1 != nil || l.err {
+		return &ParseError{"", "bad ZONEMD Scheme", l}
+	}
+	rr.Scheme = uint8(i)
+
+	c.Next() // zBlank
+	l, _ = c.Next()
+	if i, err := strconv.ParseUint(l.token, 10, 8); err != nil {
+		tokenUpper := strings.ToUpper(l.token)
+		i, ok := StringToAlgorithm[tokenUpper]
```
Comment (tmthrgd) [inline scan_rr.go:869]: This is wrong. The draft specifies only two supported algorithms SHA384 with value 1 and SHA512 with value 2; that doesn't match StringToAlgorithm. The "ZONEMD Hash Algorithms" will be a separate registry to the one used by DNSSEC.

Also, the draft doesn't specific the use of the mnemonic's here at all, instead only "unsigned decimal integer" are allowed. This should just be a parse error.

File: types.go [hunk 4]
```diff
@@ -1361,6 +1361,23 @@ func (rr *CSYNC) len(off int, compression map[string]struct{}) int {
 	return l
 }
 
+// ZONEMD RR, from draft-ietf-dnsop-dns-zone-digest
+type ZONEMD struct {
+	Hdr    RR_Header
+	Serial uint32
+	Scheme uint8
```
Comment (tmthrgd) [inline types.go:1368]: There should probably be constants for this and Hash.

**Review by tmthrgd — CHANGES_REQUESTED** [review]

File: scan_rr.go (hunk 3 above)
Comment (madestro) [reply scan_rr.go:901]: you are right, we are removing that part

File: types.go (hunk 4 above)
Comment (madestro) [reply types.go:1384]: done in #edcbb0b

File: go.mod (hunk 2 above)
Comment (eriverosr) [reply go.mod:6]: it should be fixed in #e62bf83

Comment (tmthrgd) [reply go.mod:6]: I'm still seeing the go.mod and go.sum changes in the PR diff.

File: scan_rr.go [hunk 5]
```diff
@@ -846,6 +846,70 @@ func (rr *CSYNC) parse(c *zlexer, o string) *ParseError {
 	return nil
 }
 
+func (rr *ZONEMD) parse(c *zlexer, o string) *ParseError {
+	l, _ := c.Next()
+	if (l.token == "\\#") {
```
Comment (tmthrgd) [inline scan_rr.go:851]: This is definitely not the correct place to do this nor the correct pull request to do this in, please back this out.

File: types.go [hunk 6]
```diff
@@ -148,6 +149,22 @@ const (
...
 	OpcodeNotify = 4
 	OpcodeUpdate = 5
+
+)
+
+// Used in ZONEMD https://tools.ietf.org/html/draft-ietf-dnsop-dns-zone-digest-14
+
+const (
+	// ZoneMD Accepted Schemes
+	ZoneMDSchemeReserved = 0
```
Comment (tmthrgd) [inline types.go:159]: I don't think we need the reserved constants, they're really just noise.

**Review by tmthrgd — CHANGES_REQUESTED** [review]

File: scan_rr.go (hunk 5 above)
Comment (madestro) [reply scan_rr.go:851]: where to, then?

File: types.go (hunk 6 above)
Comment (madestro) [reply types.go:159]: disagree

File: scan_rr.go (hunk 5 above)
Comment (tmthrgd) [reply scan_rr.go:851]: It looks like we only support RFC3597 in `(*ZoneParser).Next` if we don't recognise the RR type, that does seem like a bug to me. That bug needs to be fixed in the ZoneParser to work for *all* record types using the existing wire format parsing functions not here in an ad-hoc way for this specific one.

File: types.go (hunk 6 above)
Comment (tmthrgd) [reply types.go:159]: The usable constants were, the reserved ones serve no purpose and have no practical use whatsoever. It's not merely that they take up extra lines, they also appear in the documentation, require long term compatibility and impose a burden on developers understanding the code or documentation.

File: scan_rr.go (hunk 5 above)
Comment (tmthrgd) [reply scan_rr.go:851]: I've opened #1211 to add support for this.

**Review by miekg — APPROVED** [review]

**Comment by miekg** [comment]
think all requested changes have been implemented.

**Comment by miekg** [comment]
wait, where are the tests?

**Comment by eriverosr** [comment]
I've just added a test that checks if the library is able to parse the example ZONEMD RRs from [RFC 8976](https://tools.ietf.org/html/rfc8976). Tell us if you can think of any other type of tests to add.

File: parse_test.go [hunk 7]
```diff
@@ -1921,3 +1921,75 @@ func TestUnpackRRWithHeaderInvalidLengths(t *testing.T) {
...
+		rr, err := NewRR(i)
+		if err != nil {
+			t.Error("failed to parse RR: ", err)
+			continue
+		}
+		if rr.String() != o {
+			t.Errorf("`%s' should be equal to\n`%s', but is     `%s'", i, o, rr.String())
+		}
+	}
+
```
Comment (miekg) [inline parse_test.go:1994]:
```suggestion
```

**Review by miekg — APPROVED** [review]
